### PR TITLE
core: Fix handling of some invalid JPEG data (fix #4209, #8775)

### LINKED ISSUE #4209: DefineBitsJpeg2 fails when JPEG data is truncated
https://z0r.de/259 ([SWF](https://z0r.de/L/z0r-de_259.swf)) does not display anything because the `jpeg-decoder` dependency fails to decode any of the 6 embedded JPEG images.

The actual error message is: `Error running definition tag: DefineBitsJpeg2, got failed to fill whole buffer`

After extracting the binary data of these tags into separate files, every image viewer/loader I tried opens them:
 - `convert` of ImageMagick outputs a warning: `convert: Premature end of JPEG file 'image1.jpg' @ warning/jpeg.c/JPEGWarningHandler/389.`, but converts the image nevertheless
 - Gwenview shows the same result as Flash Player, and what `convert` exports
 - VS Code shows the (incomplete) bottom row of blocks as blank (transparent)

Of course Adobe Flash Player is also willing to decode these images just fine.

There is a report for this already in the repo of the `jpeg-decoder` crate, but it's closed without a proper fix: https://github.com/image-rs/jpeg-decoder/pull/98

### LINKED ISSUE #8775: Lami images bug
### Describe the bug

most of the images doesn't appear correctly
Lami's head fucking gone
where's paul
liam's white with blood

### Expected behavior

lami
lami 3
lami 2
https://www.mediafire.com/file/l8cf84j17brfopa/487956_Lami.zip/file

### Affected platform

Self-hosted version

### Operating system

Windows 10

### Browser

Mozilla Firefox

### Additional information

it's on newgrounds

### PR DESCRIPTION
**Pull request by Herschel**
1) Previously we were removing the first occurrence of the invalid 0xFFD9FFD8 byte sequence in JPEG data, but this would break the JPEG if it happened to contain this byte sequence elsewhere (for example, EXIF data). Instead, properly parse the JPEG markers searching for the invalid marker sequence. Fixes #8775.

2) Sometimes JPEGs can be lacking the final 0xFFD9 EOI marker that marks the end of the image. I suspect some JPEG optimizers might remove this to save a few bytes. Flash and most JPEG viewers still display this, but `jpeg-decoder` errors. Glue on the EOI marker if its missing to convince `jpeg-decoder` to decode these cases. Fixes #4209.

### CODE REVIEW

**Comment by torokati44** [comment]
With this new header parser, can it be easily detected if the data is too short? Then, say, extend it with zeroes to the appropriate length?
If so, that would also fix https://github.com/ruffle-rs/ruffle/issues/4209, and most likely https://github.com/ruffle-rs/ruffle/issues/2435 and https://github.com/ruffle-rs/ruffle/issues/720 too.

**Comment by Herschel** [comment]
This code explicitly bails out before the image data (SOS marker), so I'm not sure if it can help with those cases easily. I'll look into it!

**Comment by Herschel** [comment]
The z0r cases seems to simply be missing the final EOI marker, and gluing it on fixes them. I think this is safe to do, worst case the image still fails to decode.

**Review by CUB3D — APPROVED** [review]

File: render/src/utils.rs [hunk 1]
```diff
@@ -65,23 +65,79 @@ pub fn glue_tables_to_jpeg<'a>(
...
+    // See https://github.com/ruffle-rs/ruffle/issues/8775 for various examples.
+
+    // JPEG markers
+    const SOF0: u8 = 0xC0; // Start of frame
+    const RST0: u8 = 0xD0; // Restart (we shouldn't see this before SOS, but just in case)
+    const RST7: u8 = 0xD7;
+    const SOI: u8 = 0xD8; // Start of image
+    const EOI: u8 = 0xD9; // End of image
+
+    let mut data: Cow<[u8]> = if data.starts_with(&[0xFF, EOI, 0xFF, SOI]) {
```
Comment (relrelb) [inline render/src/utils.rs:90]: nit: `strip_prefix` nicely fits here.

**Comment by torokati44** [comment]
Oh, sorry, I mentioned https://github.com/ruffle-rs/ruffle/issues/2435 and https://github.com/ruffle-rs/ruffle/issues/720 above, but I misremembered, those are completely unrelated. :no_mouth:
And #4209 is indeed fixed.

File: render/src/utils.rs (hunk 1 above)
Comment (torokati44) [reply render/src/utils.rs:90]: I think once this tiny thing is addressed, this should be good to go?

Comment (torokati44) [reply render/src/utils.rs:90]: See: #9153

**Pull request merged**